### PR TITLE
GoogLeNet

### PR DESCRIPTION
**Pull request by marcinp7**

### CODE REVIEW

File: athenet/data_loader/imagenet_data_loader.py [hunk 1]
```diff
@@ -20,11 +20,12 @@ class ImageNetDataLoader(DataLoader):
     mean_rgb = [123, 117, 104]
     verbosity = 0
 
-    def __init__(self, year, buffer_size=1, train_data=True, val_data=True,
+    def __init__(self, year, image_shape, buffer_size=1, train_data=True, val_data=True,
```
Comment (glogowski-wojciech) [inline athenet/data_loader/imagenet_data_loader.py]: too long line

File: athenet/layers/__init__.py [hunk 2]
```diff
@@ -1,8 +1,12 @@
...
 from activation import *
+from concat import *
 from conv import *
 from dropout import *
 from fully_connected import *
 from norm import *
 from pool import *
 from softmax import *
+
+from inception import *
```
Comment (glogowski-wojciech) [inline athenet/layers/__init__.py]: Consider changing import * to importing specific objects. It would be safer in view of any changes in the future. Generally me @heurezjusz and @martacieslak decided to resign from using import *.

File: athenet/layers/inception.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+import numpy as np
+
+import theano
+
+from athenet.layers import Layer, ConvolutionalLayer, ReLU, MaxPool, \
+    Concatenation
+
+
+class InceptionLayer(Layer):
```
Comment (glogowski-wojciech) [inline athenet/layers/inception.py]: Why not class InceptionLayer(WeightedLayer)?

File: athenet/layers/inception.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    def __init__(self, n_filters, input_layer_name=None, name='inception'):
+        """Create inception layer.
+
+        :n_filters: List of length 6: number of filters in convolutional
+                    layers.
+        """
```
Comment (glogowski-wojciech) [inline athenet/layers/inception.py]: Insufficient comments. Consider @martacieslak comment style. It concerns all the comments.

File: athenet/data_loader/imagenet_data_loader.py (hunk 1 above)
Comment (glogowski-wojciech) [inline athenet/data_loader/imagenet_data_loader.py]: In line 65,  files = os.listdir(get_bin_path(self.val_name)) is unused

File: athenet/layers/inception.py (hunk 3 above)
Comment (marcinp7) [inline athenet/layers/inception.py]: WeightedLayer would mean that InceptionLayer has W and b properties, which is not true. Only convolutional layers inside Inception have those.

File: athenet/layers/concat.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        self.input_shapes = [layer.output_shape for layer in input_layers]
+        for input_shape in self.input_shapes[1:]:
+            if input_shape[:-1] != self.input_shapes[0][:-1]:
+                raise ValueError('all input layer image size must match')
+
+        self.input = [layer.output for layer in input_layers]
+        self.train_input = [layer.train_output for layer in input_layers]
+
+    @property
+    def output_shape(self):
```
Comment (glogowski-wojciech) [inline athenet/layers/concat.py]: Consider a comment telling what is the order of dimensions. Concerns all layers.

File: athenet/layers/conv.py [hunk 6]
```diff
@@ -93,18 +92,19 @@ def _reshape_input(self, raw_layer_input):
                                                image height, image width) or
```
Comment (glogowski-wojciech) [inline athenet/layers/conv.py]: No longer "or compatible"?

File: athenet/layers/conv.py [hunk 7]
```diff
@@ -127,8 +127,11 @@ def _get_output(self, layer_input):
         else:  # let Theano decide which implementation to use
             h, w = self.image_shape[0:2]
             pad_h, pad_w = self.padding
-            group_image_shape = (self.batch_size, n_group_channels,
-                                 h + 2*pad_h, w + 2*pad_w)
+            if self.batch_size is not None:
+                group_image_shape = (self.batch_size, n_group_channels,
+                                     h + 2*pad_h, w + 2*pad_w)
+            else:
+                group_image_shape = None
```
Comment (glogowski-wojciech) [inline athenet/layers/conv.py]: Have a look at theano.tensor.nnet.conv.conv2d: image_shape is deprecated alias for input_shape. Should be renamed.

File: athenet/layers/inception.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+from athenet.layers import Layer, ConvolutionalLayer, ReLU, MaxPool, \
+    Concatenation
+
+
+class InceptionLayer(Layer):
+    def __init__(self, n_filters, input_layer_name=None, name='inception'):
+        """Create inception layer.
+
+        :param int n_filters: List of length 6: number of filters in
```
Comment (glogowski-wojciech) [inline athenet/layers/inception.py]: I would put ":param int list of length 6 n_filters: number of filters in"

File: athenet/models/googlenet.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        InceptionLayer([128, 128, 256, 24, 64, 64], name='inception 4c'),
+        InceptionLayer([112, 144, 288, 32, 64, 64], name='inception 4d'),
+        InceptionLayer([256, 160, 320, 32, 128, 128], name='inception 4e'),
+        MaxPool(poolsize=(3, 3),
+                stride=(2, 2),
+                padding=(1, 1)),
+        InceptionLayer([256, 160, 320, 32, 128, 128], name='inception 5a'),
+        InceptionLayer([384, 192, 384, 48, 128, 128], name='inception 5b'),
+        AvgPool(poolsize=(7, 7),
+                stride=(1, 1)),
```
Comment (glogowski-wojciech) [inline athenet/models/googlenet.py]: padding=(3, 3)?

File: athenet/models/googlenet.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        if weights is None:
+            raise Exception("cannot load GoogLeNet weights")
+
+    # Normalization parameters
+    local_range = 5
+    alpha = 0.0001
+    beta = 0.75
+    k = 1
+
+    googlenet = Network([
```
Comment (glogowski-wojciech) [inline athenet/models/googlenet.py]: shadows name of function

File: athenet/models/googlenet.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+                padding=(1, 1)),
+        InceptionLayer([256, 160, 320, 32, 128, 128], name='inception 5a'),
+        InceptionLayer([384, 192, 384, 48, 128, 128], name='inception 5b'),
+        AvgPool(poolsize=(7, 7),
+                stride=(1, 1)),
+        Dropout(0.4),
+        FullyConnectedLayer(1000),
+        Softmax(),
+    ])
+    if trained:
```
Comment (glogowski-wojciech) [inline athenet/models/googlenet.py]: Maybe merge 2 "if trained:" if's. Pycharm shows warning about undeclared 'weights'.

File: athenet/layers/concat.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+from athenet.layers import Layer
+
+
+class Concatenation(Layer):
+    """Concatenation layer."""
+    def __init__(self, input_layer_names=None, name='concat'):
+        """Create concatenation layer.
+
+        :input_layer_names: List of input layers' names.
```
Comment (glogowski-wojciech) [inline athenet/layers/concat.py]: put 'param' keywords in updated files please

File: athenet/models/googlenet.py (hunk 9 above)
Comment (marcinp7) [inline athenet/models/googlenet.py]: In AveragePool? I don't think that's a good idea.

File: athenet/models/googlenet.py (hunk 11 above)
Comment (marcinp7) [inline athenet/models/googlenet.py]: I splitted that in order to not create model if weights file doesn't exist (for trained = True) - unless you have a better idea.

File: athenet/models/googlenet.py (hunk 9 above)
Comment (glogowski-wojciech) [inline athenet/models/googlenet.py]: You're right
